pkg/operator/ceph/object/controller.go

```diff
@@ -17,215 +17,329 @@ limitations under the License.
-		rookVersion: c.rookImage,
-		clusterSpec: c.clusterSpec,
-		ownerRef:    c.storeOwners(objectstore),
-		DataPathMap: cephconfig.NewStatelessDaemonDataPathMap(cephconfig.RgwType, objectstore.Name, c.clusterInfo.Name, c.dataDirHostPath),
-	}
-	if err := cfg.createOrUpdate(); err != nil {
-		logger.Errorf("failed to create or update object store %s. %v", objectstore.Name, err)
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
 	}
-}
 
-func (c *ObjectStoreController) onDelete(obj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Deleting object store for an external ceph cluster is disabled because no Ceph image is specified")
-		return
+	// validate the store settings
+	if err := validateStore(r.context, cephObjectStore); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "invalid object store %q arguments", cephObjectStore.Name)
 	}
 
-	objectstore, err := getObjectStoreObject(obj)
+	// RECONCILE SERVICE
+	logger.Debug("reconciling object store service")
+	serviceIP, err := r.reconcileService(cephObjectStore)
 	if err != nil {
-		logger.Errorf("failed to get objectstore object. %v", err)
-		return
+		cephObjectStore.Status.Phase = k8sutil.ReconcileFailedStatus
+		errStatus := opcontroller.UpdateStatus(r.client, cephObjectStore)
+		if errStatus != nil {
+			return reconcile.Result{}, errors.Wrap(errStatus, "failed to set status")
+		}
+		return reconcile.Result{}, errors.Wrap(err, "failed to reconcile service")
 	}
 
-	c.acquireOrchestrationLock()
-	defer c.releaseOrchestrationLock()
-
-	cfg := clusterConfig{context: c.context, store: *objectstore}
-	if err = cfg.deleteStore(); err != nil {
-		logger.Errorf("failed to delete object store %q. %v", objectstore.Name, err)
+	// RECONCILE POOLS
+	logger.Debug("reconciling object store pools")
+	objContext := NewContext(r.context, cephObjectStore.Name, cephObjectStore.Namespace)
+	metadataPoolToModel := *cephObjectStore.Spec.MetadataPool.ToModel("")
+	dataPoolToModel := *cephObjectStore.Spec.DataPool.ToModel("")
+	err = createPools(objContext, metadataPoolToModel, dataPoolToModel)
+	if err != nil {
+		errStatus := opcontroller.UpdateStatus(r.client, cephObjectStore)
+		if errStatus != nil {
+			return reconcile.Result{}, errors.Wrap(errStatus, "failed to set status")
+		}
+		return reconcile.Result{}, errors.Wrapf(err, "failed to create object pools")
 	}
-}
 
-// ParentClusterChanged determines wether or not a CR update has been sent
-func (c *ObjectStoreController) ParentClusterChanged(cluster cephv1.ClusterSpec, clusterInfo *daemonconfig.ClusterInfo, isUpgrade bool) {
-	c.clusterInfo = clusterInfo
-	if !isUpgrade {
-		logger.Debugf("No need to update the object store after the parent cluster changed")
-		return
+	// RECONCILE REALM
+	logger.Debug("reconciling object store realms")
+	err = reconcileRealm(objContext, serviceIP, cephObjectStore.Spec.Gateway.Port)
+	if err != nil {
```

Comment:
A lot of this error handling code is similar. What about a pattern like this?
```
if err != nil {
  return setFailedStatus(err)
}
```

Where `setFailedStatus()` attempts to update the status and if it fails it simply logs the error, and returns the `err` in a `reconcile.Result`

Reply:
Thoughts on this pattern? 
- Updating status isn't fatal to the reconcile
- 4 lines of code can be simplified to 1 in a lot of places with this error handling.